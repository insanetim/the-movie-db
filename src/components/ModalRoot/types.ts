import { ComponentProps } from 'react'

import { MODAL_COMPONENTS } from './modalComponents'

export type ModalComponentsMap = {
  [K in keyof typeof MODAL_COMPONENTS]: K
}

export type ModalProps = ComponentProps<ModalsMap[ModalTypes]>

export type ModalRootHookReturn = {
  modalProps: ModalProps | null
  modalType: ModalTypes | null
  onCancel: () => void
}

export type ModalsMap = typeof MODAL_COMPONENTS

export type ModalTypes = Extract<keyof ModalsMap, string>
