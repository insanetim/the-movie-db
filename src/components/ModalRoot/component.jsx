import useContainer from './hook'
import MODAL_COMPONENTS from './modalComponents'

const ModalRoot = () => {
  const { modalType, modalProps, onCancel } = useContainer()

  if (!modalType) {
    return null
  }

  const ModalComponent = MODAL_COMPONENTS[modalType]

  return (
    <ModalComponent
      onCancel={onCancel}
      {...modalProps}
    />
  )
}

export default ModalRoot
