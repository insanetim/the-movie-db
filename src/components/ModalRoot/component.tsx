import isNull from 'src/utils/helpers/isNull'

import useContainer from './hook'
import MODAL_COMPONENTS from './modalComponents'

const ModalRoot: React.FC = () => {
  const { modalProps, modalType, onCancel } = useContainer()

  if (isNull(modalType)) {
    return null
  }

  const ModalComponent = MODAL_COMPONENTS[modalType]

  return (
    <ModalComponent
      onCancel={onCancel}
      {...modalProps}
    />
  )
}

export default ModalRoot
