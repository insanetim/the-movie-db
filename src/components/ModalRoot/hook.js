import { useDispatch, useSelector } from 'react-redux'

import { hideModal } from 'src/store/app/actions'
import { modalPropsSelector, modalTypeSelector } from 'src/store/app/selectors'

const useContainer = () => {
  const dispatch = useDispatch()
  const modalType = useSelector(modalTypeSelector)
  const modalProps = useSelector(modalPropsSelector)

  const onCancel = () => {
    dispatch(hideModal())
  }

  return { modalType, modalProps, onCancel }
}

export default useContainer
