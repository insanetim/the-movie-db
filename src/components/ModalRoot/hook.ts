import { useAppDispatch, useAppSelector } from 'src/hooks/useRedux'
import { hideModal } from 'src/store/app/actions'
import { modalPropsSelector, modalTypeSelector } from 'src/store/app/selectors'
import type { ModalRootHook } from './types'

const useContainer = (): ModalRootHook => {
  const dispatch = useAppDispatch()
  const modalType = useAppSelector(modalTypeSelector)
  const modalProps = useAppSelector(modalPropsSelector)

  const onCancel = () => {
    dispatch(hideModal())
  }

  return { modalType, modalProps, onCancel }
}

export default useContainer
