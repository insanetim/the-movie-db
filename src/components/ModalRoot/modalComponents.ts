import ModalCreateList from './Modals/ModalCreateList'

const MODAL_COMPONENTS = {
  MODAL_CREATE_LIST: ModalCreateList,
}

export default MODAL_COMPONENTS
