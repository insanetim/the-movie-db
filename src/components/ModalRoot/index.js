import ModalRoot from './component'

export default ModalRoot
