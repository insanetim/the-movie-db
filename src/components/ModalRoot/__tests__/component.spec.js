import { render } from '@testing-library/react'

import ModalRoot from '../component'
import Wrapper from '../../../__mocks__/wrapperMock'

let mockedHookData = {
  modalType: 'CREATE_LIST_MODAL',
  modalProps: {},
  onCancel: jest.fn()
}
jest.mock('../hook', () => jest.fn(() => mockedHookData))

describe('ModalRoot component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<ModalRoot />, { wrapper: Wrapper })

    expect(asFragment()).toMatchSnapshot()
  })

  it('return null unless modalType present', () => {
    mockedHookData = {
      modalType: undefined
    }
    const { asFragment } = render(<ModalRoot />, { wrapper: Wrapper })

    expect(asFragment()).toMatchSnapshot()
  })
})
