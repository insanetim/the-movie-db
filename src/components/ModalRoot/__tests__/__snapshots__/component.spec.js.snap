// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalRoot component matches snapshot 1`] = `
<CreateListModal
  onCancel={[MockFunction]}
/>
`;

exports[`ModalRoot component return null unless modalType present 1`] = `""`;
